[FEATURE] Vérification de l'égilibité des candidats à une certif compl. dans l'espace surveillant (PIX-7833).

alexandrecoin opened the pull request:
## :unicorn: Problème

Le surveillant n’a actuellement aucun moyen d'être alerté en cas de différence entre la certification (complémentaire ou non) que le candidat est effectivement en train de passer et celle à pour laquelle il a été inscrit.

Impossible donc pour lui d’alerter à son tour le candidat ni de connaître/justifier de la nouvelle durée de certification et l’heure à laquelle doit terminer chaque candidat en se basant sur les informations de l’Espace Surveillant.

## :robot: Proposition

Vérification de l'éligibilité du candidat à la certif complémentaire qu'iel doit passer dans le usecase de récupération de la session.

## :rainbow: Remarques

On renvoie également le userId du candidat afin de déterminer si celui-ci est réconcilié ou non afin d'afficher l'avertissement.

## :100: Pour tester

Se connecter sur pix-certif avec le compte `certifsup@example.net`
Se connecter à l'espace surveillant  de la session `106420`

Dans cette session ont été ajoutés 3 candidats qui ont été réconciliés : 
- Tata (Pas de certification complémentaire)
- Titi (Inscrit à Pix+ Droit mais non éligible)
- Toto (Inscrit à Cléa et éligible)

Vérifier que seul Titi dispose d'un avertissement stipulant qu'il n'est plus éligible à la certification complémentaire.

(A réaliser en version `.fr` et `.org` avec `?lang=en`)

P-Jeremy committed:
:pencil2: api: rename complementaryCertification to enrolledComplementaryCertification

P-Jeremy committed:
:pencil2: certif: rename complementaryCertification to enrolledComplementaryCertification

P-Jeremy committed:
:sparkles: api: add userId to CertificationCandidateForSupervising

alexandrecoin committed:
:sparkles: api: check still valid badge acquisition

mariannebost committed:
:sparkles: api: add stillValidBadgeAcquisitions in CertificationCandidateForSupervising model

P-Jeremy committed:
:sparkles: api: add #isStillElligible to CertificationCandidateForSupervising

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr6137.review.pix.fr)
  * [App (.org)](https://app-pr6137.review.pix.org)
  * [Orga (.fr)](https://orga-pr6137.review.pix.fr)
  * [Orga (.org)](https://orga-pr6137.review.pix.org)
  * [Certif (.fr)](https://certif-pr6137.review.pix.fr)
  * [Certif (.org)](https://certif-pr6137.review.pix.org)
  * [Admin](https://admin-pr6137.review.pix.fr)
  * [API](https://api-pr6137.review.pix.fr/api/)
  * [Pix1D](https://pix1d-pr6137.review.pix.fr)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr6137/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr6137/environment)

mariannebost committed:
:sparkles: api: add isStillEligible to sessionForSupervising serializer

alexandrecoin committed:
:sparkles: api: add serialization for userId attribute
This information is necessary to check if the candidate has been reconciliated before the eligibility check

alexandrecoin committed:
:sparkles: certif: add warning message for non longer eligible candidates

alexandrecoin committed:
:globe_with_meridians: certif: translate eligibility warning

mariannebost commented:
Func review OK 
![Capture d’écran 2023-05-05 à 16 36 50](https://user-images.githubusercontent.com/103997660/236488799-c36401db-eb56-4666-ab5d-28d5f8a03b1a.png)
![Capture d’écran 2023-05-05 à 16 39 15](https://user-images.githubusercontent.com/103997660/236489349-0b6e89c7-813b-45a2-a749-9b55e4f8fd9c.png)

AndreiaPena commented on api/tests/tooling/domain-builder/factory/build-certification-candidate-for-supervising.js:
```diff
@@ -20,6 +21,7 @@ module.exports = function buildCertificationCandidateForSupervising({
     startDateTime,
-    complementaryCertification,
+    enrolledComplementaryCertification,
+    stillValidBadgeAcquisitions,
```
il y a ce `stillValidBadgeAcquisitions` qui pop dans le premier commit sans raison. Il permet quoi ?
update: ok je retrouve la suite dans le 4eme commit :D

AndreiaPena commented on api/tests/unit/domain/models/CertificationCandidateForSupervising_test.js:
```diff
@@ -15,4 +15,89 @@ describe('Unit | Domain | Models | Certification Candidate for supervising', fun
+      );
+
+      context(
+        'when candidate has still valid badge acquisition  not related to his complementary certification',
```
**Question :** 
Je découvre ce `stillValidBadgeAcquisitions` ? Qu'est ce qu'il fait exactement ?

Est-ce qu'il est possibile que le candidat ait une certif complémentaire différente de ce valid badge ? Si oui je veux bien une explication pour comprendre :)

alexandrecoin replied:
Le candidat peut avoir plusieurs "badges" valides.
Il faut juste vérifier que celui qu'il passe en certif corresponde à un des badges valides dans le tableau :)

AndreiaPena commented on api/lib/infrastructure/serializers/jsonapi/session-for-supervising-serializer.js:
```diff
@@ -15,6 +15,13 @@ const attributes = [
     return new Serializer('sessionForSupervising', {
+      transform(currentSessionForSupervising) {
+        currentSessionForSupervising.certificationCandidates.forEach((certificationCandidate) => {
+          certificationCandidate.isStillEligibleToComplementaryCertification =
```
**Remarques :** 
Je ne m'attendais pas à voir cette vérification ici. Pourquoi ce choix de le placer dans le serializer ?

C'est une info métier non ? ça aurait plus sa place dans le usecase.

mariannebost replied:
Par vérification tu veux parler de l'appel à la méthode isStillEligibleToComplementaryCertification() c'est bien ca ?

AndreiaPena replied:
oui :)

Steph0 replied:
Pour moi la méthode est portée par l'objet Domain CertificationCandidateForSupervising. On ne rompt pas le contrat de la Clean Archi (le serializer étant le passage Domain <-> Controller).

AndreiaPena commented on certif/app/components/session-supervising/candidate-in-list.js:
```diff
@@ -33,6 +33,20 @@ export default class CandidateInList extends Component {
+    return !this.args.candidate.isStillEligibleToComplementaryCertification;
+  }
+
+  _isReconciliated() {
```
**Question :**

On pourrait avoir à cette étape des candidats qui ne sont pas réconcilié à un utilisateur ? 

Du coup si je comprends bien, on ne montre pas le bandeau si l'élève n'est pas réconcilié à un utilisateur, mais est-ce qu'il peut passer la certif si ce n'est pas le cas ? C'est pas clair pour moi

alexandrecoin replied:
Si le candidat n'est pas encore réconcilié, il ne faut pas montrer au surveillant d'information quant à l'éligibilité ou non d'un candidat.
Une fois réconcilié, et vu que la liste des candidats est récupérée toutes les 5 secondes environ, le surveillant aura l'infomation et pourra voir si le candidat est éligible ou non, lui laissant le temps avant de commencer la session (en validant la présence du candidat) de prendre les mesures pour le prévenir du fait qu'il ne soit potentiellement pas éligible.

AndreiaPena commented on certif/app/styles/components/session-supervising/candidate-in-list.scss:
```diff
@@ -75,6 +75,10 @@
   }
+
+  &__eligibility {
+    margin-top: 4px;
```
peut-être tenter un spacing-xxs ici

AndreiaPena commented on certif/app/components/session-supervising/candidate-in-list.js:
```diff
@@ -26,13 +26,27 @@ export default class CandidateInList extends Component {
   }
 
+  get shouldDisplayNonEligibilityWarning() {
+    return this._isReconciliated() && this._isNotEligibleToEnrolledComplementaryCertification();
```
**Remarques** : je pense qu'il manque un paramètre à cet affichage, la présence d'un `enrolledComplementaryCertification` pour le candidat.

En retournant dans l'app, dans les 3 seeds à tester, Tata Tata n'a pas de certif complémentaire et donc forcément pas de message warning. 

Mais en regardant le code, il n'a pas de warning affiché uniquement parce qu'il n'est pas réconcilié mais son `isStillEligibleToComplementaryCertification` est bien à false donc si je le reconcilie, ça donne ça : 

<img width="1217" alt="Capture d’écran 2023-05-09 à 11 08 08" src="https://user-images.githubusercontent.com/58915422/237050612-06b6b4ca-66eb-403c-b52c-4042036a7e9a.png">

mariannebost replied:
Bien vu ! J'ai rajouté le candidate.enrolledComplementaryCertification au check du _isNotEligibleToEnrolledComplementaryCertification() et le message ne s'affiche plus pour les candidats pas non inscrit
